### PR TITLE
Antoine modi - maintenance generator

### PR DESCRIPTION
**Pull request by marota**
Added a new mainteance generator in grid2op/Chronics/GridStateFromFileWithForecasts.py, by defining a child class GridStateFromFileWithForecastsWithMaintenance

### CODE REVIEW

File: grid2op/Chronics/GridStateFromFileWithForecasts.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 import numpy as np
 import pandas as pd
 from datetime import timedelta
+import random
```
Comment (Tezirg) [inline grid2op/Chronics/GridStateFromFileWithForecasts.py:14]: !

File: grid2op/data_test/ieee118_R2subgrid_wcci_test_maintenance/config.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from grid2op.Backend import PandaPowerBackend
+
+config = {
+    "backend": PandaPowerBackend,
+    "action_class": TopologyAndDispatchAction,
+    "observation_class": None,
+    "reward_class": RedispReward,
+    "gamerules_class": DefaultRules,
+    "chronics_class": Multifolder,
+    "grid_value_class": GridStateFromFileWithForecasts,
```
Comment (Tezirg) [inline grid2op/data_test/ieee118_R2subgrid_wcci_test_maintenance/config.py]: `"grid_value_class": GridStateFromFileWithForecastsWithMaintenance`

**Review by BDonnot — APPROVED** [review]

File: grid2op/Chronics/GridStateFromFileWithForecasts.py (hunk 1 above)
Comment (Tezirg) [reply grid2op/Chronics/GridStateFromFileWithForecasts.py:14]: /!\

File: grid2op/tests/test_ChronicsHandler.py [hunk 3]
```diff
@@ -616,5 +618,172 @@ def test_load_still(self):
...
+        param.NO_OVERFLOW_DISCONNECTION = True
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore")
+            with make(os.path.join(PATH_DATA_TEST, "ieee118_R2subgrid_wcci_test_maintenance_3"),
+                      param=param) as env:
+                env.seed(0)
+                obs = env.reset()
+                maint = env.chronics_handler.real_data.data.maintenance
+
+                env.seed(0)
```
Comment (Tezirg) [inline grid2op/tests/test_ChronicsHandler.py:782]: Duplicate, env should be seeded once

File: grid2op/Chronics/GSFFWFWM.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    def initialize(self, order_backend_loads, order_backend_prods, order_backend_lines, order_backend_subs,
+                   names_chronics_to_backend=None):
+
+        self.name_line = order_backend_lines
+
+        # properties of maintenance
+        # self.maintenance_duration= 8*(self.time_interval.total_seconds()*60*60)#8h, 9am to 5pm
+        # 8h furation, 9am to 5pm
+        with open(os.path.join(self.path, "maintenance_meta.json"), "r", encoding="utf-8") as f:
```
Comment (Tezirg) [inline grid2op/Chronics/GSFFWFWM.py:52]: This is not robust to missing `maintenance_meta.json` (which will be missing in most datasets)

File: grid2op/Chronics/GSFFWFWM.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                if (n_Generated_Maintenance > maxDailyMaintenance):
+                    # we pick up only maxDailyMaintenance elements
+                    not_chosen = self.space_prng.choice(n_Generated_Maintenance,
+                                                        replace=False,
+                                                        size=n_Generated_Maintenance - maxDailyMaintenance)
+                    are_lines_in_maintenance[np.where(are_lines_in_maintenance)[0][not_chosen]] = False
+                maintenance_me[selected_rows_beg:selected_rows_end, are_lines_in_maintenance] = 1.0
+
+                # handle last iteration
+                n_max = int( min(res.shape[0], (nb_day_since_beg + 1) * nb_rows) / (nb_day_since_beg + 1))
```
Comment (Tezirg) [inline grid2op/Chronics/GSFFWFWM.py:145]: My brain hurts, but i've seen the TODO at the top of the loop so it's fine for now

File: grid2op/Chronics/GSFFWFWM.py (hunk 4 above)
Comment (BDonnot) [reply grid2op/Chronics/GSFFWFWM.py:52]: THis class should not be used with other datasets ;-) But i agree this could be done differently... Though handly corner cases is probably something we'll have to cover in 0.9.1 :-/

File: grid2op/tests/test_ChronicsHandler.py (hunk 3 above)
Comment (BDonnot) [reply grid2op/tests/test_ChronicsHandler.py:782]: In theory I completely agree.

However it's precisely to test that if you seed twice you get the same results. So i need to seed two time :-(

File: grid2op/Chronics/GSFFWFWM.py (hunk 5 above)
Comment (BDonnot) [reply grid2op/Chronics/GSFFWFWM.py:145]: You have not had to fix the original implementation :see_no_evil: 
Anyway best i can do in a reasonable amount of time :-/

But yes this has to be vectorized and optimized and refactor
